src/test/java/io/vertx/test/redis/CachingRedisClientTest.java

```diff
@@ -0,0 +1,125 @@
+    client = Redis.createClient(rule.vertx(), new RedisOptions().setConnectionString("redis://" + container.getHost() + ":" + container.getFirstMappedPort()));
+    cachingClient = CachingRedisClient.create(rule.vertx(), client);
+    cachingClient.connect().onComplete(onConnect -> {
+      should.assertTrue(onConnect.succeeded());
+      should.assertEquals(context, rule.vertx().getOrCreateContext());
+      api = RedisAPI.api(cachingClient);
+      before.complete();
+    });
+  }
+
+  @After
+  public void after() {
+    cachingClient.close();
+  }
+
+  @Test
+  public void testSimpleSetAndGet(TestContext should) {
+    final Async test = should.async();
+    final String key = UUID.randomUUID().toString();
+
+    api
+      .send(Command.SET, key, "hello, world")
+      .onComplete(setResponse -> {
+        should.assertTrue(setResponse.succeeded());
+
+        api
+          .get(key)
+          .onComplete(get1Response -> {
+            should.assertTrue(get1Response.succeeded());
+
+            final Response response = get1Response.result();
+
+            should.assertEquals(response.toString(), "hello, world");
+
+            // The cache impl stores the actual object, so we can assert it's from cache by
+            // asserting a subsequent get returns the same object
+            api
+              .get(key)
+              .onComplete(get2Response -> {
+                should.assertTrue(get2Response.succeeded());
+                should.assertEquals(get2Response.result().toString(), "hello, world");
+                should.assertTrue(get2Response.result() == response);
+                test.complete();
+              });
+          });
+      });
+  }
+
+  @Test
+  public void testInvalidation(TestContext should) {
+    final Async test = should.async();
+    final String key = UUID.randomUUID().toString();
+
+    cachingClient.invalidationHandler(keys -> {
+      should.assertEquals(keys.size(), 1);
+      test.complete();
+    });
+
+    api.send(Command.SET, key, "hello, world").onComplete(set1Response -> {
+      should.assertTrue(set1Response.succeeded());
+
+      api.get(key).onComplete(get1Response -> {
+        should.assertTrue(get1Response.succeeded());
+
+        final Response response = get1Response.result();
+
+        should.assertEquals(response.toString(), "hello, world");
+
+        // The cache impl stores the actual object, so we can assert it's from cache by
+        // asserting a subsequent get returns the same object
+        api.get(key).onComplete(get2Response -> {
+          should.assertTrue(get2Response.succeeded());
+          should.assertEquals(get2Response.result().toString(), "hello, world");
+          should.assertTrue(get2Response.result() == response);
+
+          api.send(Command.SET, key, "new value").onComplete(set2Response -> {
+            should.assertTrue(set2Response.succeeded());
+          });
+        });
+      });
+    });
+  }
+}
```

Comment:
Could you add a `preservesContext` test, similarly to other Redis client tests?

Reply:
Added in `CachingRedisTest` and found a few bugs! :D